lib/translator/read.cc

```diff
@@ -86,18 +86,25 @@ StatusCode LegacyRead(nvme::GenericQueueEntryCmd& nvme_cmd,
       .psdt = 0,  // PRPs are used for data transfer
       .nsid = nsid};
 
-  nvme_cmd.dptr.prp.prp1 = allocation.data_addr;
-
   alloc_len = transfer_length * lba_size;
 
+  // checks beforehand if enough space has been allocated for the buffer
+  if (buffer_in.size() < alloc_len) {
+    DebugLog("Not enough memory allocated for Read buffer");
+    DebugLog("Memory required: %d", alloc_len);
+    return StatusCode::kFailure;
+  }
+
+  nvme_cmd.dptr.prp.prp1 = reinterpret_cast<uint64_t>(buffer_in.data());
```

Comment:
This is where the prp pointer is assigned the address of the memory to where the buffer is pointing towards. I switched the logic to check the size of the buffer beforehand rather than checking it in the ReadToScsi function as that is more efficient since no call will be made to the device in case the buffer size is incorrect.